src/metabase/models/cloud_migration.clj

```diff
@@ -0,0 +1,213 @@
+
+(doto :model/CloudMigration
+  (derive :metabase/model)
+  (derive :hook/timestamped?))
+
+(methodical/defmethod t2/table-name :model/CloudMigration [_model] :cloud_migration)
+
+(t2/deftransforms :model/CloudMigration
+  {:state mi/transform-keyword})
+
+(defsetting metabase-store-migration-url
+  (deferred-tru "Store URL for migrations. Internal test use only.")
+  :visibility :internal
+  :default    "https://store-api.metabase.com/api/v2/migration"
+  :doc        false
+  :export?    false)
+
+(defsetting read-only-mode
+  (deferred-tru
+    (str "Boolean indicating whether a Metabase's is in read-only mode with regards to its app db. "
+         "Will take up to 1m to propagate to other Metabase instances in a cluster."
+         "Audit tables are excluded from read-only-mode mode."))
+  :type       :boolean
+  :visibility :admin
+  :default    false
+  :doc        false
+  :export?    false)
+
+(def ^:private read-only-mode-exceptions
+  #{;; Migrations need to update their own state
+    (t2/table-name :model/CloudMigration)
+    (t2/table-name :model/Setting)
+
+    ;; Users need to login, make queries, and we need need to audit them.
+    (t2/table-name :model/User)
+    (t2/table-name :model/Session)
+    (t2/table-name :model/LoginHistory)
+    (t2/table-name :model/AuditLog)
+    (t2/table-name :model/QueryExecution)
+    (t2/table-name :model/ViewLog)})
+
+(def ^:private ^:dynamic
+  *ignore-read-only-mode*
+  "Used during dump-to-h2, since rotate-encryption-key! over the dump will hit read-only-mode."
+  nil)
+
+;; Block write calls to most tables in read-only mode.
+(methodical/defmethod t2.pipeline/build :before [#_query-type     :toucan.statement-type/DML
+                                                 #_model          :default
+                                                 #_resolved-query :default]
+  [_query-type model _parsed-args resolved-query]
+  (when (and (read-only-mode)
+             (not (read-only-mode-exceptions (t2/table-name model)))
+             (not *ignore-read-only-mode*))
+    (throw (ex-info (tru "Metabase is in read-only-mode mode!")
+                    {:status-code 403})))
+  resolved-query)
+
+
+;; Helpers
+
+(def terminal-states
+  "Cloud migration states that are terminal."
+  #{:done :error :cancelled})
+
+(defn- cluster?
+  []
+  (>= (t2/count 'QRTZ_SCHEDULER_STATE) 2))
+
+(defn- progress-file-input-stream
+  "File input stream that calls on-percent-progress with current read progress as int from 0 to 100.
+  Does not call on-percent-progress for the same value twice. "
+  [^File file on-percent-progress]
+  (let [input-stream             (io/input-stream file)
+        length                   (.length file)
+        *bytes                   (atom 0)
+        on-percent-progress-memo (memoize on-percent-progress) ;; memoize so we don't repeat calls
+        add-bytes                #(on-percent-progress-memo (int (* 100 (/ (swap! *bytes + %) length))))
```

Comment:
I'm not sure this on-percent is doing what we think it is doing.

```clojure
cloud-migration=> (let [values (atom [])]
                    (spit "/tmp/input" #_(apply str (repeat 5 "x")) "file contents")
                    (with-open [in (progress-file-input-stream (io/file "/tmp/input") (fn [x] (swap! values conj x)))]
                      (loop []
                        (let [c (.read in)]
                          (when (pos? c)
                            (recur)))))
                    @values)
[784 1592 2423 3200 3446 4207 5061 5907 6800 7576 8423 9315 10200 100]
```

The docstring makes me thing this will send back a progression of `[4 10 ... 96 100]` type numbers, but it's quite different in reality.

Reply:
it does seem to work correctly for the following:

```clojure
  (into [] (map (fn [size]
                 (let [values (atom [])]
                   (spit "/tmp/input" "file contents")
                   (with-open [in (progress-file-input-stream (io/file "/tmp/input") (fn [x] (swap! values conj x)))]
                     (.read in (byte-array size)))
                   @values)))
        (range 0 17 3))
```

Where i feed it progressively bigger byte arrays to copy into.

Reply:
Had a bug in `progress-file-input-stream`, expected a char to be returned from arity 1 `.read` but it returns an int. Fixed.